### ISSUE TITLE
Cocoon build: Error: Null safety features are disabled for this library.

### ISSUE THREAD

**Issue by keyonghan** [flutter/flutter#67319]
Cocoon deployment is based on latest dev channel.

When building via: [flutter pub run build_runner build --release --output build --delete-conflicting-outputs ](https://github.com/flutter/cocoon/blob/master/cloud_build/app_build.sh#L12), exited with below error:

```
Step #2: [SEVERE] build_web_compilers:entrypoint on web/benchmarks.dart:
Step #2: Dart2Js finished with:
Step #2: 
Step #2: packages/dart_internal/extract_type_arguments.dart:25:7:
Step #2: Error: Null safety features are disabled for this library.
Step #2: Object? extractIterableTypeArgument(
Step #2:       ^
Step #2: packages/dart_internal/extract_type_arguments.dart:26:34:
Step #2: Error: Null safety features are disabled for this library.
Step #2:         Iterable iterable, Object? Function<T>() extract) =>
Step #2:                                  ^
Step #2: packages/dart_internal/extract_type_arguments.dart:43:7:
Step #2: Error: Null safety features are disabled for this library.
Step #2: Object? extractMapTypeArguments(Map map, Object? Function<K, V>() extract) =>
Step #2:       ^
Step #2: packages/dart_internal/extract_type_arguments.dart:43:48:
Step #2: Error: Null safety features are disabled for this library.
Step #2: Object? extractMapTypeArguments(Map map, Object? Function<K, V>() extract) =>
Step #2:                                                ^
Step #2: Error: Compilation failed.
```

**Comment by leafpetersen** [comment]
cc @jakemac53 @natebosch

**Comment by jakemac53** [comment]
What version of the Dart SDK as well as `build_web_compilers` package is this using? It would also be good to confirm the `dart_internal` package version.

**Comment by jakemac53** [comment]
Assuming this is the package in question (https://github.com/flutter/cocoon/blob/master/app/pubspec.yaml) - it looks like this is pinning `build_web_compilers` version `2.7.2` which does not pass the full `package_config.json` file to dart2js, it only passes a `.packages` file.

The `package_config` package will then give it a language version which is older than what it should have, so it gets opted out.

Can you update `build_web_compilers` to version `2.10.0` or higher?

**Comment by keyonghan** [comment]
Thanks @jakemac53 . The build passes after updating `buid_web_comilers` to `2.10.0`!

**Comment by keyonghan** [comment]
@jakemac53 @natebosch Updating `build_web_comilers` seems to cause the issue below, and it further causes the flutter benchmark page not rendered (from console: there is a `js nosuchmethod exception` in `benchmarks.dart.js`). 

Do you have any idea about this issue? Or anyone else you think can help? Thank you!

Now we are using `buid_web_comilers: 2.10.0`, and `Flutter 1.23.0-7.0.pre`, `Engine â€¢ revision 3a73d073c8`, `Tools â€¢ Dart 2.11.0 (build 2.11.0-161.0.dev).

```
Step #2: [INFO] build_web_compilers:entrypoint on web/benchmarks.dart:Running dart2js with --dump-info --fast-startup --minify --omit-implicit-checks --trust-primitives --no-source-maps --packages=.dart_tool/package_config.json -oweb/benchmarks.dart.js web/benchmarks.dart
Step #2: 
Step #2: [INFO] 2m 0s elapsed, 5254/5255 actions completed.
Step #2: [INFO] 2m 1s elapsed, 5254/5255 actions completed.
Step #2: [INFO] 2m 2s elapsed, 5254/5255 actions completed.
Step #2: [INFO] 2m 3s elapsed, 5254/5255 actions completed.
Step #2: [INFO] 2m 4s elapsed, 5254/5255 actions completed.
Step #2: [INFO] 2m 5s elapsed, 5254/5255 actions completed.
Step #2: [INFO] 2m 6s elapsed, 5254/5255 actions completed.
Step #2: [INFO] 2m 7s elapsed, 5254/5255 actions completed.
Step #2: [INFO] 2m 8s elapsed, 5254/5255 actions completed.
Step #2: [INFO] 2m 10s elapsed, 5254/5255 actions completed.
Step #2: [INFO] 2m 11s elapsed, 5254/5255 actions completed.
Step #2: [INFO] 2m 12s elapsed, 5254/5255 actions completed.
Step #2: [INFO] 2m 13s elapsed, 5254/5255 actions completed.
Step #2: [WARNING] No actions completed for 15.0s, waiting on:
Step #2:   - build_web_compilers:entrypoint on web/benchmarks.dart
Step #2: 
Step #2: [INFO] 2m 14s elapsed, 5254/5255 actions completed.
Step #2: [INFO] 2m 15s elapsed, 5254/5255 actions completed.
Step #2: [INFO] 2m 16s elapsed, 5254/5255 actions completed.
Step #2: [INFO] 2m 17s elapsed, 5254/5255 actions completed.
Step #2: [INFO] build_web_compilers:entrypoint on web/benchmarks.dart:Dart2Js finished with:
Step #2: 
Step #2: Hint: 3 hint(s) suppressed in package:source_span.
Step #2: Hint: 1 hint(s) suppressed in package:angular_components.
Step #2: Info: View the dumped .info.json file at https://dart-lang.github.io/dump-info-visualizer
Step #2: Compiled 13,369,884 characters Dart to 496,955 characters JavaScript in 19.7 seconds
Step #2: Dart file web/benchmarks.dart compiled to JavaScript: web/benchmarks.dart.js
```

**Comment by jakemac53** [comment]
It sounds like you are getting a runtime exception now, which would likely be a dart2js issue? I don't know if that could have just happened to coincide with the build_web_compilers update or something else.

I need a bit more info about the failure though to help diagnose the problem - I don't see any failures in the build logs.

**Comment by keyonghan** [comment]
This is all the building log
[app_build.log](https://github.com/flutter/flutter/files/5343013/app_build.log)

Please let me know if you need anything else, or how I can help to get you more info.

**Comment by jakemac53** [comment]
The logs appear to be cut off - possibly it is timing out?

**Comment by jakemac53** [comment]
We did switch off of using dart2js in batch mode - this stopped working with the switch to using package config files, and wasn't ever intended to be used publicly from what I understand. But that could be regressing the build time performance here which might explain the timeout if that is indeed the issue.

cc @sigmundch

**Comment by keyonghan** [comment]
Oh, sorry for the confusion, I manually terminated the build after seeing the warning reproduced. Here is the "complete" log: 
[complete_app_build.log](https://github.com/flutter/flutter/files/5343162/complete_app_build.log)

**Comment by jakemac53** [comment]
Ok, so there definitely aren't any build errors. Those warnings are ignorable - any individual action taking over 15s will log warnings and this can definitely happen if you are building a lot of dart2js apps, which is what you are seeing.

So, I don't actually understand what the problem you are experiencing is?

**Comment by keyonghan** [comment]
The problem is after the above cloud build, the rendered `benchmark.html` page failed to load with below console error. For your info. here is the dashboard based on this build: https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.html
And this the the success page: https://flutter-dashboard.appspot.com/benchmarks.html

```
benchmarks.dart.js:10445 EXCEPTION: NoSuchMethodError: method not found: 'gea' on null
STACKTRACE: 
TypeError: Cannot read property 'gea' of null
    at fP.aC (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:10524:6)
    at fP.C (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:14112:60)
    at fP.au (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:10479:13)
    at mE.C (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:14343:5)
    at iQ.au (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:10570:10)
    at iQ.on (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:10571:30)
    at nl.$0 (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:10247:50)
    at oe.$0 (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:10437:11)
    at q5.$0 (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:10725:14)
    at ks.ub [as $1$4] (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:1913:9)
REASON: null

$0 @ benchmarks.dart.js:10445
benchmarks.dart.js:7532 Uncaught TypeError: Cannot read property 'gea' of null
    at fP.aC (benchmarks.dart.js:10524)
    at fP.C (benchmarks.dart.js:14112)
    at fP.au (benchmarks.dart.js:10479)
    at mE.C (benchmarks.dart.js:14343)
    at iQ.au (benchmarks.dart.js:10570)
    at iQ.on (benchmarks.dart.js:10571)
    at nl.$0 (benchmarks.dart.js:10247)
    at oe.$0 (benchmarks.dart.js:10437)
    at q5.$0 (benchmarks.dart.js:10725)
    at ks.ub [as $1$4] (benchmarks.dart.js:1913)benchmarks.dart.js:10445 EXCEPTION: NoSuchMethodError: method not found: 'gea' on null
STACKTRACE: 
TypeError: Cannot read property 'gea' of null
    at fP.aC (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:10524:6)
    at fP.C (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:14112:60)
    at fP.au (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:10479:13)
    at mE.C (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:14343:5)
    at iQ.au (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:10570:10)
    at iQ.on (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:10571:30)
    at nl.$0 (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:10247:50)
    at oe.$0 (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:10437:11)
    at q5.$0 (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:10725:14)
    at ks.ub [as $1$4] (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:1913:9)
REASON: null

$0 @ benchmarks.dart.js:10445
benchmarks.dart.js:7532 Uncaught TypeError: Cannot read property 'gea' of null
    at fP.aC (benchmarks.dart.js:10524)
    at fP.C (benchmarks.dart.js:14112)
    at fP.au (benchmarks.dart.js:10479)
    at mE.C (benchmarks.dart.js:14343)
    at iQ.au (benchmarks.dart.js:10570)
    at iQ.on (benchmarks.dart.js:10571)
    at nl.$0 (benchmarks.dart.js:10247)
    at oe.$0 (benchmarks.dart.js:10437)
    at q5.$0 (benchmarks.dart.js:10725)
    at ks.ub [as $1$4] (benchmarks.dart.js:1913)benchmarks.dart.js:10445 EXCEPTION: NoSuchMethodError: method not found: 'gea' on null
STACKTRACE: 
TypeError: Cannot read property 'gea' of null
    at fP.aC (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:10524:6)
    at fP.C (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:14112:60)
    at fP.au (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:10479:13)
    at mE.C (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:14343:5)
    at iQ.au (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:10570:10)
    at iQ.on (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:10571:30)
    at nl.$0 (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:10247:50)
    at oe.$0 (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:10437:11)
    at q5.$0 (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:10725:14)
    at ks.ub [as $1$4] (https://version-5f30303-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:1913:9)
REASON: null

$0 @ benchmarks.dart.js:10445
benchmarks.dart.js:7532 Uncaught TypeError: Cannot read property 'gea' of null
    at fP.aC (benchmarks.dart.js:10524)
    at fP.C (benchmarks.dart.js:14112)
    at fP.au (benchmarks.dart.js:10479)
    at mE.C (benchmarks.dart.js:14343)
    at iQ.au (benchmarks.dart.js:10570)
    at iQ.on (benchmarks.dart.js:10571)
    at nl.$0 (benchmarks.dart.js:10247)
    at oe.$0 (benchmarks.dart.js:10437)
    at q5.$0 (benchmarks.dart.js:10725)
    at ks.ub [as $1$4] (benchmarks.dart.js:1913)
```

Another finding: I did a local build, and it succeeded. Here is the complete log:
[local_success_build.log](https://github.com/flutter/flutter/files/5343239/local_success_build.log)

**Comment by keyonghan** [comment]
I am not sure how the console error is related to the build warning, or not related at all... 
BTW: to make it a fair comparison, I am using the same version of flutter, engine, and dart in local as those in the cloud build.

**Comment by jakemac53** [comment]
Can you remove the `minify` argument here https://github.com/flutter/cocoon/blob/master/app/build.yaml#L15 and retry?

**Comment by keyonghan** [comment]
Removing `minify`, here is the build log:
[remove_minify.log](https://github.com/flutter/flutter/files/5343428/remove_minify.log)

The benchmark console gives (https://version-a80272c-dot-flutter-dashboard.appspot.com/benchmarks.html):
```
benchmarks.dart.js:21792 EXCEPTION: NoSuchMethodError: method not found: 'get$usesStyleEncapsulation' on null
STACKTRACE: 
TypeError: Cannot read property 'get$usesStyleEncapsulation' of null
    at ViewBenchmarkGrid0.initViewRoot$1 (https://version-a80272c-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:21960:18)
    at ViewBenchmarkGrid0.build$0 (https://version-a80272c-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:28387:34)
    at ViewBenchmarkGrid0.create$2 (https://version-a80272c-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:21861:19)
    at _ViewBenchmarkGridHost0.build$0 (https://version-a80272c-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:28785:25)
    at ComponentFactory.create$2 (https://version-a80272c-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:22063:23)
    at ComponentFactory.create$1 (https://version-a80272c-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:22066:19)
    at ApplicationRef_bootstrap_closure.call$0 (https://version-a80272c-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:21394:43)
    at ChangeDetectionHost_run_closure.call$0 (https://version-a80272c-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:21779:33)
    at NgZone__run_closure.call$0 (https://version-a80272c-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:22384:26)
    at StaticClosure._rootRun [as call$1$4] (https://version-a80272c-dot-flutter-dashboard.appspot.com/benchmarks.dart.js:4370:16)
REASON: null

call$0 @ benchmarks.dart.js:21792
benchmarks.dart.js:15574 Uncaught TypeError: Cannot read property 'get$usesStyleEncapsulation' of null
    at ViewBenchmarkGrid0.initViewRoot$1 (benchmarks.dart.js:21960)
    at ViewBenchmarkGrid0.build$0 (benchmarks.dart.js:28387)
    at ViewBenchmarkGrid0.create$2 (benchmarks.dart.js:21861)
    at _ViewBenchmarkGridHost0.build$0 (benchmarks.dart.js:28785)
    at ComponentFactory.create$2 (benchmarks.dart.js:22063)
    at ComponentFactory.create$1 (benchmarks.dart.js:22066)
    at ApplicationRef_bootstrap_closure.call$0 (benchmarks.dart.js:21394)
    at ChangeDetectionHost_run_closure.call$0 (benchmarks.dart.js:21779)
    at NgZone__run_closure.call$0 (benchmarks.dart.js:22384)
    at StaticClosure._rootRun [as call$1$4] (benchmarks.dart.js:4370)
```

**Comment by jakemac53** [comment]
This isn't something I can really diagnose I don't think - it sounds like either an Angular related thing or dart2js related thing... hard to tell.

Possibly some bad interaction with the new sdk and/or new library deps.

**Comment by matanlurey** [comment]
I investigated for a moment. The code is quite diverged from what we use internally, but it appears the code that initializes the CSS has been completely tree-shaken/removed from resulting Dart2JS (It's missing a call to initialize component styles), which is normally null guarded. I suspect some sort of null-safety regression/interaction.

Our own JS goldens have code that looks like:

```js
if (styles == null)
        styles = $.ViewGoldenComponent0__componentStyles = O._UnscopedComponentStyles$(C.List_empty, null);
```

But the `benchmarks.dart.js` has none of this.

**Comment by sigmundch** [comment]
There was recently a bad interaction of this sort which affected `is` tests (see for example https://dart-review.googlesource.com/c/sdk/+/162247/), but I'm not aware yet if there are other issues of this kind with `== null` checks.

The change landed a month ago (https://github.com/dart-lang/sdk/commit/8d68480746dbb7878bdc6e3926d0324760af6c5f) and I believe it is part of the 2.11.0-161.0.dev build that seems that you are using above, so I'm guessing there is something else we need to look into.

@fishythefish @rakudrama - could one of you take closer look?

**Comment by fishythefish** [comment]
I can try to take a look, but I'm a little out of my depth here - what code do I need to build and how do I build it with SDK `master`?

**Comment by sigmundch** [comment]
I just remembered that we recently had a bug in how we interpreted the default values of the null-safety flags, and as a result we were generating invalid code. The bug was around for about 5 days and was part of a few dev releases until it was detected.

@keyonghan - could you let us know if this issue persists in the latest Dart beta release? In particular one newer than 2.11.0-180.0.dev?

My understanding is that precisely in Dart SDK 2.11.0-161.0.dev the issue was introduced (https://github.com/dart-lang/sdk/commit/fed7fcc0c816d69973fd78baacec24db29647ddc), and the fix went out first in 2.11.0-180.0.dev (https://github.com/dart-lang/sdk/commit/cea8843939891bb3cca717a25b5f081ba1c9b9ef)

**Comment by keyonghan** [comment]
@sigmundch I just tried building based on latest framework flutter `master` (Flutter 1.23.0-8.0.pre.223, Framework revision 82ff246156 (2 hours ago), Engine revision 2eac514f26, Tools Dart 2.11.0 (build 2.11.0-186.0.dev), it succeeded without console error on the dashboard!

I guess the problem is indeed fixed in the latest version.

Since we are relying on flutter framework `dev` channel rather than `master`, we may just need to wait for latest changes merged to `dev`.

**Comment by godofredoc** [comment]
@keyonghan  please mark the other issue as duplicate of this one.

**Comment by keyonghan** [comment]
Closing this one this the fix has been available on tot. Will leave https://github.com/flutter/flutter/issues/67457 open to track further.

**Comment by sigmundch** [comment]
excellent! thanks for verifying!
